File: 21558 - Nithin/week 3/1696-jump-game-vi/1696-jump-game-vi.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> dp(n);
        dp[0] = nums[0];
        deque<int> queue{0};
        
        for(int i = 1; i < n; i++){
            if(queue.front() < i - k)
                queue.pop_front();
            dp[i] = nums[i] + dp[queue.front()];
            while(!queue.empty() && dp[queue.back()] <= dp[i])
                queue.pop_back();
            queue.push_back(i);
        }
         return dp.back();
    }
};

File: 21553 - Sai Nandan D/Week 3/1696_Jump Game VI.py
from collections import deque

class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        dp = [nums[0]]
        q = deque([0])        
        
        for i in range(1, len(nums)):
            dp.append(dp[q[0]] + nums[i])
            while q and dp[i] > dp[q[-1]]:
                q.pop()
            q.append(i)
            if q[0] == i-k: q.popleft()
                
        return dp[-1]
